pkg/manager/member/tikv_scaler.go

```diff
@@ -52,176 +53,230 @@ func (s *tikvScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet
-			safeTimeDeadline := pod.CreationTimestamp.Add(5 * s.deps.CLIConfig.ResyncDuration)
-			if time.Now().Before(safeTimeDeadline) {
-				// Wait for 5 resync periods to ensure that the following situation does not occur:
-				//
-				// The tikv pod starts for a while, but has not synced its status, and then the pod becomes not ready.
-				// Here we wait for 5 resync periods to ensure that the status of this tikv pod has been synced.
-				// After this period of time, if there is still no information about this tikv in TidbCluster status,
-				// then we can be sure that this tikv has never been added to the tidb cluster.
-				// So we can scale in this tikv pod safely.
-				resetReplicas(newSet, oldSet)
-				return fmt.Errorf("TiKV %s/%s is not ready, wait for 5 resync periods to sync its status", ns, podName)
-			}
-		}
-
-		pvcs, err := util.ResolvePVCFromPod(pod, s.deps.PVCLister)
-		if err != nil {
-			return fmt.Errorf("tikvScaler.ScaleIn: failed to get pvcs for pod %s/%s in tc %s/%s, error: %s", ns, pod.Name, ns, tcName, err)
-		}
-		for _, pvc := range pvcs {
-			if err := addDeferDeletingAnnoToPVC(tc, pvc, s.deps.PVCControl); err != nil {
-				return err
-			}
+	var (
+		errs                         []error
+		finishedOrdinals             = sets.NewInt32()
+		updateReplicasAndDeleteSlots bool
+	)
+	for _, ordinal := range ordinals {
+		finished, err := scaleInOne(ordinal)
+		if !finished {
+			errs = append(errs, err)
+		} else {
+			finishedOrdinals.Insert(ordinal)
+			updateReplicasAndDeleteSlots = true
 		}
+	}
 
-		setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
-		return nil
+	if updateReplicasAndDeleteSlots {
+		setReplicasAndDeleteSlotsByFinished(-1, newSet, oldSet, ordinals, finishedOrdinals)
+	} else {
+		resetReplicas(newSet, oldSet)
 	}
-	return fmt.Errorf("TiKV %s/%s not found in cluster", ns, podName)
+	return errorutils.NewAggregate(errs)
 }
 
-func (s *tikvScaler) preCheckUpStores(tc *v1alpha1.TidbCluster, podName string) (bool, error) {
+func (s *tikvScaler) preCheckUpStores(tc *v1alpha1.TidbCluster, podName string, upTikvStoreCount, deleteStoreCount, maxReplicas int) bool {
 	if !tc.TiKVBootStrapped() {
 		klog.Infof("TiKV of Cluster %s/%s is not bootstrapped yet, skip pre check when scale in TiKV", tc.Namespace, tc.Name)
-		return true, nil
+		return true
 	}
 
-	pdClient := controller.GetPDClient(s.deps.PDControl, tc)
-	// get the number of stores whose state is up
-	upNumber := 0
-
-	storesInfo, err := pdClient.GetStores()
-	if err != nil {
-		return false, fmt.Errorf("failed to get stores info in TidbCluster %s/%s", tc.GetNamespace(), tc.GetName())
-	}
-	// filter out TiFlash
-	for _, store := range storesInfo.Stores {
-		if store.Store != nil {
-			if store.Store.StateName == v1alpha1.TiKVStateUp && util.MatchLabelFromStoreLabels(store.Store.Labels, label.TiKVLabelVal) {
-				upNumber++
-			}
-		}
-	}
+	// decrease deleted store in this round.
+	upNumber := upTikvStoreCount - deleteStoreCount
```

Comment:
What if the previously deleted stores are not up?

Reply:
Yeah, increase `deleteStoreCount` after `pdClient.DeleteStore()` right now, but may not in `Up` state. I will add a check.